compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1418 @@
+      new LocalEnv(params.zip(args).toMap, ddef.symbol, outer)(valsMap = mutable.Map.empty, varsMap = mutable.Map.empty)
+
+    def setLocalVal(x: Symbol, value: Value)(using data: Data, ctx: Context): Unit =
+      assert(!x.isOneOf(Flags.Param | Flags.Mutable), "Only local immutable variable allowed")
+      data match
+      case localEnv: LocalEnv =>
+        assert(!localEnv.vals.contains(x), "Already initialized local " + x.show)
+        localEnv.vals(x) = value
+      case _ =>
+        throw new RuntimeException("Incorrect local environment for initializing " + x.show)
+
+    def setLocalVar(x: Symbol, addr: Heap.Addr)(using data: Data, ctx: Context): Unit =
+      assert(x.is(Flags.Mutable, butNot = Flags.Param), "Only local mutable variable allowed")
+      data match
+      case localEnv: LocalEnv =>
+        assert(!localEnv.vars.contains(x), "Already initialized local " + x.show)
+        localEnv.vars(x) = addr
+      case _ =>
+        throw new RuntimeException("Incorrect local environment for initializing " + x.show)
+
+    /**
+     * Resolve the environment owned by the given method.
+     *
+     * The method could be located in outer scope with intermixed classes between its definition
+     * site and usage site.
+     *
+     * Due to widening, the corresponding environment might not exist. As a result reading the local
+     * variable will return `Cold` and it's forbidden to write to the local variable.
+     *
+     * @param meth  The method which owns the environment
+     * @param thisV The value for `this` of the enclosing class where the local variable is referenced.
+     * @param env   The local environment where the local variable is referenced.
+     *
+     * @return the environment and value for `this` owned by the given method.
+     */
+    def resolveEnv(meth: Symbol, thisV: Value, env: Data)(using Context): Option[(Value, Data)] = log("Resolving env for " + meth.show + ", this = " + thisV.show + ", env = " + env.show, printer) {
+      env match
+      case localEnv: LocalEnv =>
+        if localEnv.meth == meth then Some(thisV -> env)
+        else resolveEnv(meth, thisV, localEnv.outer)
+      case NoEnv =>
+        // TODO: handle RefSet
+        thisV match
+        case ref: OfClass =>
+          resolveEnv(meth, ref.outer, ref.env)
+        case _ =>
+          None
+    }
+
+    def withEnv[T](env: Data)(fn: Data ?=> T): T = fn(using env)
+  end Env
+
+  /** Abstract heap for mutable fields
+   */
+  object Heap:
+    abstract class Addr:
+      /** The static object which owns the mutable slot */
+      def owner: ClassSymbol
+
+    /** The address for mutable fields of objects. */
+    private case class FieldAddr(regions: Regions.Data, field: Symbol, owner: ClassSymbol) extends Addr
+
+    /** The address for mutable local variables . */
+    private case class LocalVarAddr(regions: Regions.Data, sym: Symbol, owner: ClassSymbol) extends Addr
+
```

Comment:
Abstract heap addresses can refer to mutable fields or local variables. A heap address also stores other necessary information, such as its owner and the region it belongs to. The value corresponds to the address is the join of all mutations to this address.